Adding support for "All", "Favorites" and "Last Played" systems

pjft opened the pull request:
Summary:
- Refactoring System Environment data
- Added Virtual System Manager class
- Added "all", "favorites" and "last played" systems
- Added GuiInfoPopup class for notifications
- Added Favorites to metadata, as well as a shortcut to toggle favorites

---

Long version:

This took a bit longer than I intended, as throughout the development I ended up going back and refactoring a few of the things I was doing. As such, it's a bit bigger than I'd have liked it to be.

Several recent topics in the forum for the last month have been around creating custom systems - not only "favorites", but also theme-based systems. Truth is, the reason why we call the ES top-level entries "systems" is purely because ES itself was designed around it. I feel there's a use case for creating entries at the top-level in ES that don't necessarily have to map 1:1 to an actual "system" per se.

This first PR adds the option for users to enable 3 such entries - which I'm naming "Virtual Systems", for lack of better terminology. These are:
- An entry with all games in your system;
- An entry with the games you last played (effectively, a list of games with play count larger than 1, and sorted by last played time, descending)
- An entry for all games you marked as "Favorite". This was pending a separate PR, but as that went through "design paralysis due to undefined scope", I ended up implementing it here with the original submitters' blessing. All credit goes to @zigurana on his work on the ES-Kids mode, and to @sokiee for pushing for it in the main branch recently!

These entries are disabled by default, but users can choose to enable any of them. These require specific folders in the theme, otherwise they'll be unthemed. I'm working with Udb23 to create these in Carbon, so that the default theme supports these. Themes that currently support it are the Comic Book theme and the Steampunk one, who I asked to add those entries. 

Other notes:
- I decoupled the "runtime/execution" settings from the SystemData class as, truth be told, they shouldn't belong there but rather to the File that's executed. There's now a "SystemEnvironmentData" structure that contains the runtime environment for the current systems, but the FileData class contains it. It's also the FileData class that now has the method to launch (rather than the system). This makes more semantic sense, as you're launching a file, not a system.
- Since the favorites implementation stayed away from any theme- or visual-related UI changes, users can add a game to the favorites system by editing the metadata, or - if the system is enabled - by pressing the "Y" button. Given the discussions taking place on how to best proceed, I created a popup notification that gives the user feedback on whether they added or removed an item to the favorites list. This keyboard shortcut may be temporary, but for now I feel it helps.
- There may be a couple of stubs or "TO DO" entries in the VirtualSystemsManager class, as I'm considering expanding this to allow users to create their own custom entries/games collections and edit/manage those from inside ES, provided that their themes have such folders.

Looking forward to testing, feedback, the usual. 

Even though this is an option (and disabled by default), I wouldn't necessarily support this being merged before Carbon is updated, as few people will be able to use it.

Thanks.

zigurana commented:
For now, I am respond with a simple : Wow!

Looking forward to going through these changes later tonight!

Op ma 26 jun. 2017 18:27 schreef pjft <notifications@github.com>:

> Summary:
>
>    - Refactoring System Environment data
>    - Added Virtual System Manager class
>    - Added "all", "favorites" and "last played" systems
>    - Added GuiInfoPopup class for notifications
>    - Added Favorites to metadata, as well as a shortcut to toggle
>    favorites
>
> ------------------------------
>
> Long version:
>
> This took a bit longer than I intended, as throughout the development I
> ended up going back and refactoring a few of the things I was doing. As
> such, it's a bit bigger than I'd have liked it to be.
>
> Several recent topics in the forum for the last month have been around
> creating custom systems - not only "favorites", but also theme-based
> systems. Truth is, the reason why we call the ES top-level entries
> "systems" is purely because ES itself was designed around it. I feel
> there's a use case for creating entries at the top-level in ES that don't
> necessarily have to map 1:1 to an actual "system" per se.
>
> This first PR adds the option for users to enable 3 such entries - which
> I'm naming "Virtual Systems", for lack of better terminology. These are:
>
>    - An entry with all games in your system;
>    - An entry with the games you last played (effectively, a list of
>    games with play count larger than 1, and sorted by last played time,
>    descending)
>    - An entry for all games you marked as "Favorite". This was pending a
>    separate PR, but as that went through "design paralysis due to undefined
>    scope", I ended up implementing it here with the original submitters'
>    blessing. All credit goes to @zigurana <https://github.com/zigurana>
>    on his work on the ES-Kids mode, and to @sokiee
>    <https://github.com/sokiee> for pushing for it in the main branch
>    recently!
>
> These entries are disabled by default, but users can choose to enable any
> of them. These require specific folders in the theme, otherwise they'll be
> unthemed. I'm working with Udb23 to create these in Carbon, so that the
> default theme supports these. Themes that currently support it are the
> Comic Book theme and the Steampunk one, who I asked to add those entries.
>
> Other notes:
>
>    - I decoupled the "runtime/execution" settings from the SystemData
>    class as, truth be told, they shouldn't belong there but rather to the File
>    that's executed. There's now a "SystemEnvironmentData" structure that
>    contains the runtime environment for the current systems, but the FileData
>    class contains it. It's also the FileData class that now has the method to
>    launch (rather than the system). This makes more semantic sense, as you're
>    launching a file, not a system.
>    - Since the favorites implementation stayed away from any theme- or
>    visual-related UI changes, users can add a game to the favorites system by
>    editing the metadata, or - if the system is enabled - by pressing the "Y"
>    button. Given the discussions taking place on how to best proceed, I
>    created a popup notification that gives the user feedback on whether they
>    added or removed an item to the favorites list. This keyboard shortcut may
>    be temporary, but for now I feel it helps.
>    - There may be a couple of stubs or "TO DO" entries in the
>    VirtualSystemsManager class, as I'm considering expanding this to allow
>    users to create their own custom entries/games collections and edit/manage
>    those from inside ES, provided that their themes have such folders.
>
> Looking forward to testing, feedback, the usual.
>
> Even though this is an option (and disabled by default), I wouldn't
> necessarily support this being merged before Carbon is updated, as few
> people will be able to use it.
>
> Thanks.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/RetroPie/EmulationStation/pull/168
> Commit Summary
>
>    - Adding support for Virtual Systems
>
> File Changes
>
>    - *M* es-app/CMakeLists.txt
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-0>
>    (6)
>    - *M* es-app/src/FileData.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-1>
>    (161)
>    - *M* es-app/src/FileData.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-2>
>    (26)
>    - *M* es-app/src/FileFilterIndex.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-3>
>    (3)
>    - *M* es-app/src/FileSorts.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-4>
>    (9)
>    - *M* es-app/src/Gamelist.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-5>
>    (4)
>    - *M* es-app/src/MetaData.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-6>
>    (1)
>    - *M* es-app/src/MetaData.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-7>
>    (1)
>    - *M* es-app/src/SystemData.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-8>
>    (151)
>    - *M* es-app/src/SystemData.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-9>
>    (31)
>    - *M* es-app/src/SystemScreenSaver.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-10>
>    (33)
>    - *A* es-app/src/VirtualSystemManager.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-11>
>    (319)
>    - *A* es-app/src/VirtualSystemManager.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-12>
>    (68)
>    - *M* es-app/src/guis/GuiGamelistOptions.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-13>
>    (9)
>    - *A* es-app/src/guis/GuiInfoPopup.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-14>
>    (109)
>    - *A* es-app/src/guis/GuiInfoPopup.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-15>
>    (24)
>    - *M* es-app/src/guis/GuiMenu.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-16>
>    (9)
>    - *M* es-app/src/guis/GuiMenu.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-17>
>    (1)
>    - *M* es-app/src/guis/GuiMetaDataEd.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-18>
>    (11)
>    - *A* es-app/src/guis/GuiVirtualSystemsOptions.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-19>
>    (99)
>    - *A* es-app/src/guis/GuiVirtualSystemsOptions.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-20>
>    (32)
>    - *M* es-app/src/views/ViewController.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-21>
>    (22)
>    - *M* es-app/src/views/ViewController.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-22>
>    (1)
>    - *M* es-app/src/views/gamelist/BasicGameListView.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-23>
>    (9)
>    - *M* es-app/src/views/gamelist/BasicGameListView.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-24>
>    (2)
>    - *M* es-app/src/views/gamelist/IGameListView.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-25>
>    (2)
>    - *M* es-app/src/views/gamelist/ISimpleGameListView.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-26>
>    (29)
>    - *M* es-core/src/Settings.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-27>
>    (1)
>    - *M* es-core/src/Util.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-28>
>    (81)
>    - *M* es-core/src/Util.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-29>
>    (7)
>    - *M* es-core/src/Window.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-30>
>    (13)
>    - *M* es-core/src/Window.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-31>
>    (2)
>    - *M* es-core/src/components/SwitchComponent.cpp
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-32>
>    (17)
>    - *M* es-core/src/components/SwitchComponent.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-33>
>    (2)
>    - *M* es-core/src/resources/Font.h
>    <https://github.com/RetroPie/EmulationStation/pull/168/files#diff-34>
>    (1)
>
> Patch Links:
>
>    - https://github.com/RetroPie/EmulationStation/pull/168.patch
>    - https://github.com/RetroPie/EmulationStation/pull/168.diff
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/168>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AF0i2MHw8aGZCcfghlahvf2Fi7GKWeU5ks5sH9vXgaJpZM4OFjDR>
> .
>

meleu commented:
FYI: I compiled and executed this branch on my x86 and everything runs pretty fine.

Will test on my raspi1 in a couple of days.

pjft commented:
Thanks all.

@zigurana I'm certainly looking forward to having you and others go through the changes in more detail, if/when you have the time. This is a lot to go through.

lilbud over at the forums caught a small bug which I'm fixing as we speak, so I hope to update the PR today or tomorrow morning. It's about the notification popup showing up after coming back from launching a game - the reason for that being quite simple, and related to the fact that we re-init SDL, which means the SDL ticks re-start from zero :)

Other than that, should be functionally solid, I hope.

Darknior commented:
Same ... WOAW i must try it quickly lol
Thanks a lot for this hard work 👍 
Have you add now FAVORITE like a FLAG on the gamelist.xml ?
I don't really understand why you are speaking about a subfolder for the theme ?
Is it possible to add the FLAG for Hidden games too ? Some times it can be cool to hide a BIOS file or a file we don't want to use a game like a DOS BAT or EXE file. Thanks

pjft commented:
Hi all. I just updated the PR with a fix for the bug lilbud reported.

@Darknior : 
- There's a favorite tag in gamelist.xml, under the game element. Just like in ES Kids. Correct.
- These are [top-level entries in the ES carousel](https://www.youtube.com/watch?v=rUvua2yPIyQ), so they need to be themed. I should add a guard for the user, when enabling those entries, warning them if the theme doesn't have those folders. I'll do that.
- A flag for hiding entries is not within the scope of this change.

pjft commented:
Added a warning when the user is enabling the systems, in case the theme doesn't support them.

zigurana commented on es-app/src/VirtualSystemManager.cpp:
```diff
@@ -0,0 +1,415 @@
+			FileFilterIndex* index = newSys->getIndex();
+			for(auto sysIt = SystemData::sSystemVector.begin(); sysIt != SystemData::sSystemVector.end(); sysIt++)
+			{
+				if ((*sysIt)->getName() != "retropie") {
```
maybe use an exclude list for systems and games that we want to exclude? Hardcoding the exceptions seems prone to errors in the future.

pjft replied:
Sure. It may make it a tad slower (iterating a list and compare with each) but sounds fair.

I take it you mean, still, a centralized definition of the exclusions (but still in the code), rather than a user-selectable options-based systems-exclusion list?

pjft replied:
Done. Since we actually use this "==retropie" comparison in other places (namely on the "go to random game" function) I added a "isGameSystem" function to SystemData, that - for now - just checks whether it's equal to retropie, but can be extended in the future. String comparison is still cheaper than a list comparison, so if we know we only have a single comparison for now, let's keep it like that.

zigurana commented on es-app/src/VirtualSystemManager.cpp:
```diff
@@ -0,0 +1,415 @@
+	}
+}
+
+void VirtualSystemManager::loadCustomVirtualSystems()
```
What is the plan for this stub?

pjft replied:
Custom user collections :)

2-player best of. Fighting games. Platform games. Mario games. Whatever.

It will be phase two.

zigurana commented on es-app/src/VirtualSystemManager.cpp:
```diff
@@ -0,0 +1,415 @@
+						}
+
+						if (include) {
+							FileData* newGame = new FileData(*gameIt, newSys);
```
So this is essentially a duplication of something that is already known. How bad will this affect performance, especially in the case of AUTO_ALL_GAMES, for people with very large libraries?

pjft replied:
Hm. Do you mean creating the new FileData as a whole, rather than re-using a pointer to it?

It is a fair comment on the large library angle. As I said, I benchmarked (time only, not memory, which is a concern as well) loading the virtual systems/collections for my 4k game collection (400 something on last played and favorites, I believe), and the time it takes reports as less than 1 second (I'm getting the milliseconds and dividing by 1000). 

I could, for the sake of fun, run a test to create these systems 10 times simulating a 40k game collection, and see how memory and time goes.

I suppose, if memory and performance is ok, I'd rather keep it given that the underlying architecture (views-controller-model) expects the 1:1 relationship of SystemData to FileData to be upheld. I'm not sure I'm keen on fidgeting with the views so that whenever they try to get the respective SystemData, to have to figure out what exact collection they're referring to. But it's something to consider, for sure.

pjft replied:
A non-scientific benchmark (few times tried, not enough to get a scientific average - also, it may depend on other resources). I just measured the amount of memory used and time to load on the Pi. Not ideal, but a reasonable, not-completely-accurate proxy for what to expect.

```
No Virtual Systems (so, what I expect it to be in the current build):
pre-loading-ES: 63MB used
post-loading-ES: 153MB

Collections - loading the games 100x to simulate a large library (5k games * 100 = 500k games loaded in new FileDatas)
pre-ES starting: 63MB memory used
post-ES-load: 538MB memory used (+385MB, give or take)
time taken to build systems: 79038ms (79s for 100 times / 500k games loaded)

Collections - loading 10x (5k games * 10 = 50k games loaded in new FileDatas)
pre-ES starting: 63MB memory used
post-ES-load: 189MB memory used (+35MB give or take)
time taken to build systems: 7908ms (7.9s for 10 times / 50k games loaded)

Collections - normal (5k extra games in FileDatas)
pre: 63MB used
post-load: 157MB used (+5MB give or take)
time taken to build systems: 792ms (0.79s for 1 time / 5k games)
```

With a library of 50k games we should be talking of an increase of 8secs in loading time, and about 35MB in RAM being used.

If loading 5k games, we're talking about <1sec added loading time, and about 5MB extra memory.

I'd vouch for keeping it as it is, given the architectural expectations between views, system data and file data, and the effort of refactoring that vs the actual resources being used here, but I'm always open to pushbacks.

zigurana commented on es-app/src/guis/GuiInfoPopup.cpp:
```diff
@@ -0,0 +1,114 @@
+	}
+}
+
+bool GuiInfoPopup::updateState()
```
There is already some animation code in es-core/src/animations/ do we need this here as well?
I must admit I never got that to work reliably myself, so maybe it's not worth bothering.

pjft replied:
I know :( 

I struggled with it for a while, and in the end I didn't get the NinePatchComponent to fade in and out using Lambda. It's probably because it's using the image Alpha, rather than using the Opacity setting - and I imagine that the LambdaAnimation is using Opacity.

That's why - if you read the updateState method - I need to change the Opacity for one of them, and the color alpha for the other one.

It's terrible.

I could consider redefining the SetOpacity method for the NinePatchComponent and see if that would solve that.

zigurana commented on es-app/src/guis/GuiVirtualSystemsOptions.cpp:
```diff
@@ -0,0 +1,118 @@
+
+	std::vector<std::string> selSystems = autoOptionList->getSelectedObjects();
+	std::string out = "";
+	// from a vector of system names get comma separated string
```
maybe move this out in a separate method in Util.cpp?

pjft replied:
Will look into it!

pjft replied:
Done?

zigurana commented on es-app/src/VirtualSystemManager.cpp:
```diff
@@ -0,0 +1,415 @@
+	return FileSorts::SortTypes.at(0);
+}
+
+std::vector<std::string> VirtualSystemManager::stringToSystems(std::string systemString)
```
Maybe move this method to Util.cpp?

pjft replied:
LGTM. I think it no longer accesses any instance data.

pjft replied:
Done.

pjft commented:
Hi @zigurana ! Thank you very much for the detailed review of the code and architecture! It's very much appreciated :)

You brought up a lot of things, all of them valid, so let me try to address them as well as possible - and later in the day or tomorrow I'll actually go into the code comments you left.

**New features!**

_I've always thought of this concept as collections, sets, or containers._

Agreed. Same here. I think "SystemData" and "FileData" as a concept differ significantly from what their class names actually suggest.

_Additionally, you've introduced the favorites metadata type, together with the MD_Bool type and the small extension of the SwitchComponent. Thanks for picking that up, another one of my dropped balls, I'm afraid._

Not at all - you had it very much done on your branch, I am terribly sorry that there's no reasonable way to bring over your code with attribution. The only reason I added it here is because the other favorites PR didn't materialize, so instead of launching "all" and "last played" for now, I just did it all. Thanks for putting it together!

_Finally you've introduced a new popup window GUI element, allowing us to shoot short messages at the user, which I'm sure will come in very handy at some later point in time. I'd love to add images though!_

No objections - I think it's a good idea. It is a grid component. I might add it now, or we may extend it later, no concerns. Probably later when we have actual images or icons to put there (info/warning/something?).

**Naming**

_Now to be bold; I ask you to consider renaming the virtual systems to 'collections'. (...)_

You'd be surprised :) Before I started this a few weeks ago I had fully drafted a bug here on GitHub to give everyone the heads up and debate the options around this. I then decided not to because it would feel extremely self-assured from my end, and we'd also probably get stuck in design debates at the time. :)

I initially wanted to rename SystemData and FileData as a whole, but decided against it because it would break a lot of people's work, and also would make importing code from other branches a tremendous pain. So that was quickly scrapped.

As for renaming the new class, I have no objections whatsoever. "Virtual Systems" are not the best name whatsoever, and are probably more "jargony" than they should. I'm more than happy to make it "Collections". Even just yesterday I was going to name the menu "Custom Systems", as the users don't need to know what a "Virtual System" is -- but you're right, they don't even need to know what a "System" is in the first place. 

Consider it approved.

_I see a couple of instances of code like if (mVirtual) { //do awesome stuff!}, and I wonder if we would not be better off with a new subclass for FileData, like VirtualFileData, where you just override these particular methods with the new behaviour you want._ 

Yeah... This does make sense. I think that's how I tried to do it (actually, with the SystemData class) but ended up failing miserably. Opposite yo your self-description in the intro, I am a Software Engineer by training, and I did it for a while professionally, but haven't really been involved in hands on development of anything serious for the last... 10 years? This has been a nice creative outlet for me, but I'm rustier than I'd like to.

I'll think about the best approach there and will try to do it via a subclass. It is how I initially devised it, I just struggled with the implementation and then said "screw it, let's get something working first".

_In any case, I am not completely sure why we would need to effectively duplicate all the FileData objects? (At least for those that are part of a collection). Could we not make the FileData have a 'regular' SystemData parent, as well as (multiple) additional CollectionData parents?_

We _could_. However, right now, the data architecture - from gamelists to populating the systems in memory - is System-driven. Populating a system gamelist involves getting a system's/collection's data, and iterating it once to populate. So, more than having a FileData have additional parents, we need the parents to have the right FileData as children.

I could use a shared pointer - i.e. having the "collections" re-use the FileData objects in memory. For some reason I considered that briefly but decided not to. I suppose it's something related to how SystemData expects a 1:1 bi-directional relationship with its FileData entries, and how the Views then also expect a 1:1 relationship with a dedicated SystemData. There are way too many "fileData->getSystem()" dependencies in the View portion of EmulationStation that I didn't want to tangle with as it's just prone to errors.

I honestly didn't do a proper memory benchmarking - only on the "time to load these systems", which was consistently under 1 second for my 4000+ games collection - to see if creating dedicated FileData entries for these systems had a significant impact on the system. My gut feeling suggests "no", as most of these entries have either pointers to existing objects in memory, or store plain data types such as strings and such. But I have no hard evidence of that.

I could easily get a high-level memory read before and after enabling those systems, though.

But this way I managed to keep the same elegant/lazy/streamlined expected relationship between all these entities.

Summary: unless there's significant objection to this, or it has a material impact on the resource consumption, I'd probably leave it for now. But I'm very much open to feedback here, and happy to revisit

**SystemData class extension**

_The same reasoning could be followed here regarding subclassing, but there is very little mIsVirtualSystem specific code here. Still, I feel that subclassing is cleaner, especially because they are different concepts (collections vs systems)._

Might try it. It's a fair request. Once again, same reason for the FileData subclassing - my own failure and limitations rather than a deliberate decision.

**String handling**

_Like I said, good job in collecting some of these methods into Util.cpp. You have introduced some new string to vector and vise-versa methods though, and I wonder if they could not be placed here also. They might come in handy for the definition and storage of filter types, tags etc._

Sounds fair. I'll look into which ones aren't object specific and move then there!

----

All in all, thanks for the feedback and comments.

I'm keen on hearing back from you and others on this whole PR, to make sure we make the right decisions here.

I'll comment on your code suggestions later in the day or tomorrow - my week is somewhat packed, and today in particular I likely won't have the time to go through them with the time they deserve. :)

Best!

meleu commented:
FYI, this branch is running smoothly on my raspi1.

This virtual systems feature is like a dream becoming true. Pretty handy! :-)

joolswills commented:
Thanks for your work on this. Not had a chance to try myself yet, but looks good.

joolswills commented on es-app/src/MetaData.cpp:
```diff
@@ -19,6 +19,7 @@ MetaDataDecl gameDecls[] = {
 	{"publisher",	MD_STRING,				"unknown",			false,		"publisher",			"enter game publisher"},
```
Happy to use space indentation for laying out stuff like this (but keep the tabs for now for the general code indentation).

pjft replied:
LGTM.

pjft replied:
Done.

pjft commented on es-app/src/FileData.cpp:
```diff
@@ -1,62 +1,44 @@
+	  mSourceFileData(file->getMetadataSourceFileData()),
+	  mParent(NULL),
+	  metadata(file->getMetadataSourceFileData()->metadata),
+	  mIsVirtual(true)
```
LGTM (Looks good to me).

pjft replied:
Done.

pjft commented on es-app/src/FileData.cpp:
```diff
@@ -134,12 +126,27 @@ std::vector<FileData*> FileData::getFilesRecursive(unsigned int typeMask, bool d
+		return getShortKey();
+}
+
+FileData* FileData::getMetadataSourceFileData()
```
Hm. We will return the actual file data, or the source file data (i.e. the original filedata of which the current instance is a clone of). Would "get()" not be misleading? Happy to reconsider the name, though - it's just that "get()" might not be accurate.

zigurana replied:
From a 'client' point of view, would you care if the FileData you are getting is the original source, or a copy of the source? For example, are we expecting to have changes in the copy that are not propagated through to the original? Will we maybe at somepoint need a 'regular' `get()` as well?

pjft replied:
Well... Your comment actually makes sense. From a client point of view, you are right. They don't care. They want to get the original instance of the File being manipulated - use cases are, usually, for metadata editing, for file deletion, etc.

In that sense, I wouldn't be shocked with renaming it to something more transparent. I just associate "get()" with "getting the actual instance". Would "getSource()" be better? "getFileData()" (given that, for FileData objects it returns "this", but for "CollectionFileData" objects it returns the FileData object it's cloned from)?

Thanks!

pjft commented on es-app/src/FileData.cpp:
```diff
@@ -185,3 +192,67 @@ void FileData::sort(const SortType& type)
+		mVirtualName = removeParenthesis(mSourceFileData->metadata.get("name"));
+		boost::trim(mVirtualName);
+		mVirtualName += " [" + mSourceFileData->getSystem()->getName() + "]";
+	}
```
The virtual name is used as the display name for Virtual entries. We are _appending_ the system name as this is what the user sees in the list, so that in a multi-system collection the user can identify what games are from which system, and so that the user can play the version of the game they want - think of simple clashes like Alladin SNES/MegaDrive, but even games that showed up in more systems than I can count. 

I imagine one could also have a theme element per entry to mark it as belonging to a specific system, but that's beyond the scope (and design discussion) of this current change - similar to how to represent Favorites in the UI.

See:

https://youtu.be/rUvua2yPIyQ?t=10

zigurana replied:
Thanks for explaining.

pjft commented on es-app/src/FileData.h:
```diff
@@ -57,12 +56,21 @@ class FileData
+
+	std::string getKey();
+	inline std::string getLongKey() { return getPath().string(); };
+	inline std::string getShortKey() { return getPath().filename().string(); };
```
LGTM.

pjft replied:
Done (though doesn't seem to be updated here?).

pjft commented on es-app/src/FileSorts.cpp:
```diff
@@ -36,8 +36,9 @@ namespace FileSorts
-		std::string name2 = file2->getName();
+		// we compare the actual metadata name, as virtual files have the system appended which messes up the order
+		std::string name1 = file1->metadata.get("name");
+		std::string name2 = file2->metadata.get("name");
```
LGTM. We were no longer comparing filenames a long time ago, though. We are comparing File names :)

But yes, I can certainly change it.

pjft replied:
Done.

pjft commented on es-app/src/FileSorts.cpp:
```diff
@@ -62,9 +63,11 @@ namespace FileSorts
 	{
 		//only games have lastplayed metadata
+		// since it's stored as a POSIX string (YYYYMMDDTHHMMSS,fffffffff), we can compare as a string
+		// as it's a lot faster than the time casts and then time comparisons
```
The comment is there to clarify to anyone why this change was done. Thanks!

pjft commented on es-app/src/SystemData.cpp:
```diff
@@ -17,45 +15,43 @@ std::vector<SystemData*> SystemData::sSystemVector;
 	mName = name;
 	mFullName = fullName;
-	mStartPath = startPath;
+	mEnvData = envData;
```
LGTM!

pjft commented on es-app/src/views/gamelist/ISimpleGameListView.cpp:
```diff
@@ -129,6 +132,32 @@ bool ISimpleGameListView::input(InputConfig* config, Input input)
+		{
+			if(Settings::getInstance()->getString("VirtualSystemsAuto").find("favorites") != std::string::npos)
+			{
+				FileData* cursor = getCursor();
```
LGTM. In fact, this won't be "toggle favorites" for long. It will be "add game to collection (game, collection)".

pjft replied:
Done for now.

pjft commented on es-app/src/views/ViewController.cpp:
```diff
@@ -339,6 +339,20 @@ std::shared_ptr<IGameListView> ViewController::getGameListView(SystemData* syste
 	return view;
 }
 
+void ViewController::deleteGameListView(SystemData* system)
```
Maybe. As I said, way too many refactorings while developing. I don't think I have any future use for this one. I did a code cleanup before committing, but this one must have slipped - thanks for catching it!

pjft replied:
Done.

pjft commented on es-app/src/guis/GuiMenu.cpp:
```diff
@@ -373,6 +377,11 @@ void GuiMenu::openScreensaverOptions() {
 
+void GuiMenu::openVirtualSystemSettings() {
+	GuiVirtualSystemsOptions* ggf = new GuiVirtualSystemsOptions(mWindow);
+	mWindow->pushGui(ggf);
```
LGTM :) Actually, it was done like this for consistency with previous options, but you're right - ggf is a pure copy and paste of the previous method :)

I can certainly consolidate it in a single line.

pjft replied:
Done.

pjft commented:
Thanks all, and thanks once again @zigurana for the thoughtful comments. 

I will set out to implement the majority of the changes, as they're quite valid comments.

I'll update when things are running stable again. It'll might end up taking me 7-10 days, though as usual I hope to get this out sooner. 

Thanks!

zigurana commented:
Hey @pjft , thanks for your expedient work on this. It all looks good to me now. I'll strive to testing this fully this week, when work family, and other social commitments permit.

pjft commented:
Not at all, thanks for all the feedback and availability. I tried to incorporate most feedback you gave, except for the ones that I explicitly stated that I wasn't sure yet. So that's still open for review if you'd feel so inclined :)

I've squashed my new changes in a single commit (I'll squash them again if/when this is approved).

The sub-class creation gave me a bit of trouble as I started running into some segmentation faults at times, because of lost pointers or partial destructors.

I _think_ I have it sorted now, but I'd now benefit from functional testing, certainly, to make sure things are working as expected after these changes.

Most of my tests have been related to editing metadata, deleting files, and seeing how they are handled. I _think_ it's all working as intended, but you never know.

Thanks once again, and please, take your time!

zigurana commented on es-app/src/FileData.h:
```diff
@@ -7,6 +7,7 @@
 #include "MetaData.h"
 
 class SystemData;
+class SystemEnvironmentData;
```
So here SystemEnvironmentData is a class...

zigurana commented on es-app/src/SystemData.h:
```diff
@@ -8,23 +8,31 @@
 #include "FileFilterIndex.h"
+#include "CollectionSystemManager.h"
+
+struct SystemEnvironmentData
```
... but here its a struct?

pjft replied:
...would you believe me if I said I did that to mess with you? :P 

Thanks for spotting that, it's obviously an oversight from my end! Glad you caught that now, otherwise in the future someone would certainly look at that and be utterly confused.

It should be a struct.

zigurana replied:
OK, no problem. There are some other additional warnings, but none of them very interesting. Only this one seemed like one that could lead to trouble down the road.

pjft replied:
Not at all, thanks for catching that. The compiler on the Pi doesn't warn me of anything!

zigurana commented:
Ok, I'm stress testing the auto-systems now, I can get ES to crash by first adding a couple of games to the (currently unthemed) favorites system, and then removing them all while in the favorites GameListView.

The debugger breaks out at the top of FileData::getChildrenListToDisplay():
![debugging_autosystems1](https://user-images.githubusercontent.com/6103768/27930345-81fe6088-6296-11e7-99b3-67f457c2250b.png)

Somehow, the cursor does not get to be replaced by the empty placeholder, which is exactly the mechanism we have for that type of empty list situations, (see ISimpleGameListView::onFileChanged(),
That will probably need to happen in BasicGameListView::remove().

zigurana commented:
Recent games seems robust, even when the last-played date is in the future, it works fine (claiming that I played it 'just now', soit). Good!

pjft commented:
Thanks for spotting that case - the one where you remove the last one. 

I have quite a few favorites on my list, so I forgot to test that. I'll check that, thank you!

zigurana commented on es-app/src/guis/GuiInfoPopup.cpp:
```diff
@@ -0,0 +1,114 @@
+	}
+	if (mSize.y() > maxHeight) {
+		s->setSize(mSize[0], maxHeight);
+		mSize[1] = maxHeight;
```
Ok, you're going to hate me for this...

When I replace the warning text about missing auto-system theme nodes in the xml, with something ridiculously long, such as the two first paragraphs of Lorem Ipsum...., I get this:

![debugging_autosystems2](https://user-images.githubusercontent.com/6103768/27932082-aee2974e-629c-11e7-8788-2f4c43b25ea6.png)
This is on a 1024x768 windowed screen, but that should not matter, as all sizes are in relative screen sizes anyways. As you can see, the textComponent is placed centered on the screen, basically sticking out on the top and bottom. 
Now, I did not expect this to go completely right, but then again, I was not expecting this either. First of all, it overrides the top-bottom padding, second, the centering obscures the top part.

We should consider an implementation like the description box in the detailed gamelistview, which is a textComponent, within a scrollableContainerComponent.

The expected / optimal behavior for me would be a popup window, respecting the paddings, that allows a user to scroll through the text using up-down or similar.

Now, do you need to fix this now? No, probably not, seeing as the only instance of this is hard-coded by yourself. But this PR is the introduction of this new functionality, so hey, I thought, lets test it. 
I am also fine with it if you want me to make a separate issue out of it, to be fixed at a later date. 

Actually, with the refactoring of the GameListViews, there will be a ScrollableTextComponent, which might be more suitable to use, so maybe we better wait for that to land. 

(Argh! All these dependencies: lets wait for this, and this, and this, like a reverse bootstrapping, boot-undoing? Barrel full of crab?)

zigurana commented:
Ok, I'm done, for now at least :-P. 
You see I had to get rather creative to find something wrong at all with the code, really getting out into the fringes of normal-use. I think overall this works great, and will be a superb addition to ES.

Bravo!

pjft commented:
Thank you for the thorough testing!

I've actually just updated the PR with a few things:
- Changed the "Random game selection" buttons to what we discussed earlier today. Shall I get rid of the option in the menu?
- Added what I hope fixes the "remove" case - would you be kind enough to, when you do have the time, to test it out again?
- I think the Loren Ipsum test is quite valid :) That being said, I'd like you to confirm something: the GuiInfoPopup code you refer to is only used as a popup to state "Game X was added to Collection Y". You seem to be describing the warning for "hey, you theme doesn't support this", which is using a standard GuiMsgBox. I don't object in any way to changing the GuiInfoPopup to use a scrolling container if the width/height gets unmanageable, but I want to make sure that that is what you're suggesting, and not changing the GuiMsgBox component? If it's the latter, we can - and should - certainly change it, but probably in a separate PR (happy to do that at a later stage) otherwise I may end up bundling a lot of small fixes in a new functionality PR.

Let me know, and thank you very much for the thorough testing here, seriously! :)

zigurana commented:
Hi @pjft!
You are right of course, about the GuiMsgBox. I don't know what I was thinking. Lets tackle that slight (mis)behavior in a separate Issue/PR.
I've tried the same thing in the GuiInfoPopup, and then the result is:
![debugging_autosystems3](https://user-images.githubusercontent.com/6103768/27946948-428e3ece-62f4-11e7-9d64-2e1c790cfc38.png)
The text is forced to remain on a single line, which for the current use-case (a brief + not too intrusive message) is fine, but it's interesting to see that there is no automatic reflow or wrapping of the text going on.

pjft commented:
Great :)

I just added a "sort by system" option to the sorting menu. 

Hopefully this is as good as it gets! :)

Thanks for all the help and testing!

zigurana commented:
So regarding the remove functionality:
this is your latest proposal for BasicGameListView::Remove():
```
void BasicGameListView::remove(FileData *game, bool deleteFile)
{
	if (deleteFile)
		boost::filesystem::remove(game->getPath());  // actually delete the file on the filesystem
	if (getCursor() == game)                     // Select next element in list, or prev if none
	{
		std::vector<FileData*> siblings = game->getParent()->getChildrenListToDisplay();
		if (siblings.size() > 0)
		{
			auto gameIter = std::find(siblings.begin(), siblings.end(), game);
			auto gamePos = std::distance(siblings.begin(), gameIter);
			if (gameIter != siblings.end())
			{
				if ((gamePos + 1) < siblings.size())
				{
					setCursor(siblings.at(gamePos + 1));
				} else if ((gamePos - 1) > 0) {
					setCursor(siblings.at(gamePos - 1));
				}
			}
		}
		else
		{
			addPlaceholder();
		}
	}
	delete game;                                 // remove before repopulating (removes from parent)
	onFileChanged(game, FILE_REMOVED);           // update the view, with game removed
}
```
This (still) does not work (i.e. does not insert a placeholder when the list is empty, because the last sibling is only removed after the check. At first, I thought that simply adding a placeholder earlier would solve the issue (so `if (siblings.size() > 1)`. But that did not work either. This is because the onFileChanged() method calls for a valid cursor, which the placeholder is not (its added to the mList, but you cannot set the cursor to it, currently).
So yeah, still not perfect, but we are getting there!

pjft commented:
I see.

Thanks for testing!

I won't waste more of your time with this "correspondence testing" :)

I suppose I'll have to craft a library on my end where I can test it and
see if it works.

Thanks for catching it!


- Paulo Tavares

On Fri, Jul 7, 2017 at 3:25 PM, D. Polders <notifications@github.com> wrote:

> So regarding the remove functionality:
> this is your latest proposal for BasicGameListView::Remove():
>
> void BasicGameListView::remove(FileData *game, bool deleteFile)
> {
> 	if (deleteFile)
> 		boost::filesystem::remove(game->getPath());  // actually delete the file on the filesystem
> 	if (getCursor() == game)                     // Select next element in list, or prev if none
> 	{
> 		std::vector<FileData*> siblings = game->getParent()->getChildrenListToDisplay();
> 		if (siblings.size() > 0)
> 		{
> 			auto gameIter = std::find(siblings.begin(), siblings.end(), game);
> 			auto gamePos = std::distance(siblings.begin(), gameIter);
> 			if (gameIter != siblings.end())
> 			{
> 				if ((gamePos + 1) < siblings.size())
> 				{
> 					setCursor(siblings.at(gamePos + 1));
> 				} else if ((gamePos - 1) > 0) {
> 					setCursor(siblings.at(gamePos - 1));
> 				}
> 			}
> 		}
> 		else
> 		{
> 			addPlaceholder();
> 		}
> 	}
> 	delete game;                                 // remove before repopulating (removes from parent)
> 	onFileChanged(game, FILE_REMOVED);           // update the view, with game removed
> }
>
> This (still) does not work (i.e. does not insert a placeholder when the
> list is empty, because the last sibling is only removed after the check. At
> first, I thought that simply adding a placeholder earlier would solve the
> issue (so if (siblings.size() > 1). But that did not work either. This is
> because the onFileChanged() method calls for a valid cursor, which the
> placeholder is not (its added to the mList, but you cannot set the cursor
> to it, currently).
> So yeah, still not perfect, but we are getting there!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/168#issuecomment-313696927>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAV7UZPSn3gXUg3hKWygL9PQnL_BCukks5sLj-_gaJpZM4OFjDR>
> .
>

zigurana commented:
No worries, I like the puzzle!
At least, now I do, I think I got it!

``` 
void BasicGameListView::setCursor(FileData* cursor)
{
	//if (cursor->isPlaceHolder())
	//	return;
	if(!mList.setCursor(cursor) && (!cursor->isPlaceHolder()))
	{
		populateList(cursor->getParent()->getChildrenListToDisplay());
		mList.setCursor(cursor);
		...
```
This works!

pjft commented:
Huh. I wonder... But does the placeholder get added then?

Or is it just that it doesn't crash?

I was thinking of

```
        if (deleteFile)
boost::filesystem::remove(game->getPath());  // actually delete the file on
the filesystem
SystemData parent = game->getParent();
if (getCursor() == game)                     // Select next element in
list, or prev if none
{
std::vector<FileData*> siblings = parent->getChildrenListToDisplay();
auto gameIter = std::find(siblings.begin(), siblings.end(), game);
auto gamePos = std::distance(siblings.begin(), gameIter);
if (gameIter != siblings.end())
{
if ((gamePos + 1) < siblings.size())
{
setCursor(siblings.at(gamePos + 1));
} else if ((gamePos - 1) > 0) {
setCursor(siblings.at(gamePos - 1));
}
}
}
delete game;                                 // remove before repopulating
(removes from parent)
if(parent->getChildrenListToDisplay().size() == 0)
{
addPlaceholder();
}
else
{
onFileChanged(game, FILE_REMOVED);           // update the view, with game
removed
}

```

but I'll do some testing with your suggestion as well.

It just takes a lot of time to compile :)

Thanks!


- Paulo Tavares

On Fri, Jul 7, 2017 at 3:49 PM, D. Polders <notifications@github.com> wrote:

> No worries, I like the puzzle!
> At least, now I do, I think I got it!
>
> void BasicGameListView::setCursor(FileData* cursor)
> {
> 	//if (cursor->isPlaceHolder())
> 	//	return;
> 	if(!mList.setCursor(cursor) && (!cursor->isPlaceHolder()))
> 	{
> 		populateList(cursor->getParent()->getChildrenListToDisplay());
> 		mList.setCursor(cursor);
> 		...
>
> This works!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/168#issuecomment-313703567>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAV7T_qp-1LyqHQlWmHqVsxp8fsVOG9ks5sLkVygaJpZM4OFjDR>
> .
>

pjft commented:
Ok. So I tested this a bit further, and I found that both my solution
crashed, as well as yours, it seems?

I don't see the "logical" difference between what gets executed between the
previous code and your last proposal? Either way, I unfortunately managed
to make it crash as well.

The main problem is in set/getCursor, because we are attempting to retrieve
and then perform an operation on an object that has just been deleted, but
whose memory address/pointer still exists in the TextListComponent.

I am now removing the object from the list, and explicitly adding the
placeholder if it's empty - especially as a zero-sized list will fail
multiple times because of all the assertions in the code.

I think it's now fixed, but I appreciate your probing as usual :)

Thank you!


- Paulo Tavares

On Fri, Jul 7, 2017 at 3:57 PM, Paulo Tavares <paulo.j.tavares@gmail.com>
wrote:

> Huh. I wonder... But does the placeholder get added then?
>
> Or is it just that it doesn't crash?
>
> I was thinking of
>
> ```
>         if (deleteFile)
> boost::filesystem::remove(game->getPath());  // actually delete the file
> on the filesystem
> SystemData parent = game->getParent();
> if (getCursor() == game)                     // Select next element in
> list, or prev if none
> {
> std::vector<FileData*> siblings = parent->getChildrenListToDisplay();
> auto gameIter = std::find(siblings.begin(), siblings.end(), game);
> auto gamePos = std::distance(siblings.begin(), gameIter);
> if (gameIter != siblings.end())
> {
> if ((gamePos + 1) < siblings.size())
> {
> setCursor(siblings.at(gamePos + 1));
> } else if ((gamePos - 1) > 0) {
> setCursor(siblings.at(gamePos - 1));
> }
> }
> }
> delete game;                                 // remove before repopulating
> (removes from parent)
> if(parent->getChildrenListToDisplay().size() == 0)
> {
> addPlaceholder();
> }
> else
> {
> onFileChanged(game, FILE_REMOVED);           // update the view, with game
> removed
> }
>
> ```
>
> but I'll do some testing with your suggestion as well.
>
> It just takes a lot of time to compile :)
>
> Thanks!
>
>
> - Paulo Tavares
>
> On Fri, Jul 7, 2017 at 3:49 PM, D. Polders <notifications@github.com>
> wrote:
>
>> No worries, I like the puzzle!
>> At least, now I do, I think I got it!
>>
>> void BasicGameListView::setCursor(FileData* cursor)
>> {
>> 	//if (cursor->isPlaceHolder())
>> 	//	return;
>> 	if(!mList.setCursor(cursor) && (!cursor->isPlaceHolder()))
>> 	{
>> 		populateList(cursor->getParent()->getChildrenListToDisplay());
>> 		mList.setCursor(cursor);
>> 		...
>>
>> This works!
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/RetroPie/EmulationStation/pull/168#issuecomment-313703567>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AVAV7T_qp-1LyqHQlWmHqVsxp8fsVOG9ks5sLkVygaJpZM4OFjDR>
>> .
>>
>
>

zigurana commented:
Well, the difference is that `mList.setCursor(cursor)` tries to set the cursor first, and only if that fails, will we even consider the rest of the if-condition. In the original, we don't even try to set the cursor in case of a placeholder FileData*.
At least, that was my intention. I had it work, but were other ways for it to fail, apparently. I'll try your latest version now.

pjft commented:
Oh, I see. You are correct in that.

I don't object to phrasing it the way you had it. :)
On Fri, 7 Jul 2017 at 19:32, D. Polders <notifications@github.com> wrote:

> Well, the difference is that mList.setCursor(cursor) tries to set the
> cursor first, and only if that fails, will we even consider the rest of the
> if-condition. In the original, we don't even try to set the cursor in case
> of a placeholder FileData*.
> At least, that was my intention. I had it work, but were other ways for it
> to fail, apparently. I'll try your latest version now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/168#issuecomment-313759650>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAV7a3Xbys-Zk-wzZ0yD5VzwW_Gef4Wks5sLnnPgaJpZM4OFjDR>
> .
>

zigurana commented on es-app/src/CollectionSystemManager.cpp:
```diff
@@ -0,0 +1,442 @@
+
+SystemData* CollectionSystemManager::findCollectionSystem(std::string name)
+{
+	SystemData* sys;
```
`sys` is never used, I think?

zigurana commented:
@pjft It does not matter which way we use, your method works as well, slightly more lines, so I win the brevity points this round ;-).
I think with these latest (known) wrinkles straightened out, we might be ready to commit, or not?

pjft commented:
Sounds good:)

I'll just test it in the shorter version then to confirm nothing breaks,
and address your latest comment.

I hope we're good to commit then:)

Thank you very much!
On Fri, 7 Jul 2017 at 19:56, D. Polders <notifications@github.com> wrote:

> @pjft <https://github.com/pjft> It does not matter which way we use, your
> method works as well, slightly more lines, so I win the brevity points this
> round ;-).
> I think with these latest (known) wrinkles straightened out, we might be
> ready to commit, or not?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/RetroPie/EmulationStation/pull/168#issuecomment-313765118>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AVAV7bdnM8DWnJ78RNdtWDv9jzGJ--SLks5sLn9PgaJpZM4OFjDR>
> .
>

pjft commented:
Added the changes we discussed, and confirmed it compiles. I am unable to re-test those last changes at this exact moment, but reading through  it it doesn't seem as if anything drastic changes - IList.h can handle a setCursor of NULL in the worst case scenario.

A sincere thanks for the thorough review and feedback!

pjft commented:
Tested this, seems to be good to go.

joolswills commented:
Thanks for all the work on this. Merging!

pjft commented:
Thank you!
